lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5581,6 +5583,152 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
   return None;
 }
 
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known);
+
+// Given a type, check if its in 'known'. Do the same with its subtypes.
+// When the type is nominal, get its declaration and check if any if its
+// member's type is in known by calling 'checkDeclCircularityImpl'
+bool isTypePresentForCircularity(Type type, std::set<TypeBase *> &known) {
+  auto desugared = type->getDesugaredType();
+
+  if (auto NT = dyn_cast<NominalType>(desugared)) {
+    if(auto NTT = NT->getDecl()->getDeclaredTypeInContext()) {
+      if (known.find(NTT.getPointer()) != known.end()) {
+        return true;
+      }
+    }
+    // the type is cleared, now we check its member.
+    return checkDeclCircularityImpl(NT->getDecl(), known);
+  }
+
+  // check each generic argument, this takes care of parameter from declaring
+  // context as well as stuff like Optionals
+  if (auto BGT = dyn_cast<BoundGenericType>(desugared)) {
+    for (auto T: BGT->getGenericArgs()) {
+      if(isTypePresentForCircularity(T, known)) {
+        return true;
+      }
+    }
+
+    // check the overall generic type
+    auto BGTT = BGT->getDecl()->getDeclaredTypeInContext();
```

Comment:
That requires type => decl lookup each time a tuple element type or generic argument comes up. Type in the set is the lesser evil, no?